src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+                match event {
+                    ListenEvent::Progress(progress) => {
+                        assert!(
+                            timely::PartialOrder::less_equal(&self.read_progress, &progress),
+                            "{:?} PARTIAL ORDER: {:?} {:?}",
+                            self.name,
+                            self.read_progress,
+                            progress
+                        );
+                        self.read_progress = progress;
+                    }
+                    ListenEvent::Updates(updates) => {
+                        for ((_, value), timestamp, diff) in updates {
+                            let partition = value.expect("Unable to decode partition id");
+                            self.persisted_timestamp_bindings
+                                .entry(partition)
+                                .or_insert_with(VecDeque::new)
+                                .push_back((timestamp, MzOffset { offset: diff }));
+                        }
+                    }
+                }
+            }
+
+            // Decide if we're up to date enough to attempt to persist new bindings!  If not, don't bother trying and shortcut to reading in again
+            if !PartialOrder::less_equal(&self.write_upper, &self.read_progress) {
+                // Use intentionally wrong value of `expected_upper` in order to update the upper for next iteration.
+                let empty: [(((), PartitionId), Timestamp, Diff); 0] = [];
+                self.write_upper = self
+                    .write_handle
+                    .compare_and_append(
+                        empty,
+                        Antichain::from_elem(Timestamp::minimum()),
+                        Antichain::from_elem(Timestamp::minimum() + 1),
+                    )
+                    .await
+                    .expect("Timestamper upper update CAA")
+                    .with_context(|| "Timestamper upper update compare and append")?
+                    .map_err(|Upper(actual_upper)| actual_upper)
+                    .expect_err("Antichain was advanced past min on initialization");
+
+                continue;
+            }
+
+            // Compact bindings we've read in and see if we're able to assert a timestamp for any of the input offsets
+            for (partition, offset) in observed_max_offsets.iter() {
+                if matched_offsets.contains_key(partition) {
+                    continue;
+                }
+                let bindings = match self.persisted_timestamp_bindings.get_mut(&partition) {
+                    Some(bindings) => bindings,
+                    None => continue,
+                };
+                if bindings.is_empty() {
+                    continue;
+                }
+
+                // Compact as able, relying on all messages being in ascending offset order
+                while bindings.len() > 1
```

Comment:
> What guarantees that there will always be at least one binding left in the end

the loop condition ensures that neither `unwrap` in the loop will trigger -- specifically the `bindings.len() > 1`

I think your snippet is easier to reason about.  Let me try and work that in

Reply:
Unfortunately this snippet won't work for a couple reasons:

1) because `bindings` isn't guaranteed to be partitioned.  From the docs on `VecDeque::partition_point`: "If the deque is not partitioned, the returned result is unspecified and meaningless, as this method performs a kind of binary search."
2) Because the offsets are structured as diffs (not absolute offsets!), we need to do the compaction before we compare. 

I suppose another way to do this would to iterate through the bindings, keeping a running count of the sum as well as keeping track of the relevant index (via a complicated Iter::accum) -- and then break once we're at the point where we should stop compacting.  After that drain and update the first element.  Though that seems far more complicated to reason about correctness imo!